### PR TITLE
Fix TEXT_SIMILARITY ONNX inference for models without token_type_ids

### LINKED ISSUE #4924: [BUG] TEXT_SIMILARITY inference fails with Input mismatch
**What is the bug?**
TEXT_SIMILARITY inference fails with Input mismatch, looking for: [input_ids, attention_mask] when deploying the Hugging Face ONNX model onnx-community/bge-reranker-v2-m3-ONNX.

The model registers and deploys successfully in opensearch 3.6, but every inference request fails before model execution.

After reviewing the ML Commons source code, TextSimilarityTranslator always generates three input tensors (input_ids, attention_mask and token_type_ids), while the exported ONNX model only declares two inputs (input_ids and attention_mask).

It is unclear whether this is an incompatibility in TextSimilarityTranslator, a DJL binding issue, or a limitation of TEXT_SIMILARITY with ONNX models exported from Hugging Face.

**How can one reproduce the bug?**
using this  docker-compose.yml
```
services:
  opensearch:
    image: ${OPENSEARCH_IMAGE}
    container_name: opensearch
    restart: unless-stopped
    env_file:
      - .env
    environment:
      - cluster.name=opensearch-cluster
      - node.name=opensearch-node1
      - discovery.type=single-node
      - bootstrap.memory_lock=true
      - OPENSEARCH_JAVA_OPTS=${OPENSEARCH_JAVA_OPTS}
      - OPENSEARCH_INITIAL_ADMIN_PASSWORD=${OPENSEARCH_INITIAL_ADMIN_PASSWORD}
      - plugins.ml_commons.controller_enabled=true
      - TZ=${TZ}
      - OIDC_ISSUER=${OIDC_ISSUER}
      - OIDC_CONNECT_URL=${OIDC_CONNECT_URL}
      - OIDC_CLIENT_ID=${OIDC_CLIENT_ID}
      - OIDC_CLIENT_SECRET=${OIDC_CLIENT_SECRET}
      - OIDC_ROLE_ADMIN=${OIDC_ROLE_ADMIN}
      - http_proxy=${http_proxy}
      - https_proxy=${https_proxy}
      - no_proxy=${no_proxy}
      - HTTP_PROXY=${HTTP_PROXY}
      - HTTPS_PROXY=${HTTPS_PROXY}
      - NO_PROXY=${NO_PROXY}
    ulimits:
      memlock:
        soft: -1
        hard: -1
      nofile:
        soft: 65536
        hard: 65536
    volumes:
      - ./data:/usr/share/opensearch/data
      - ./securityconfig:/usr/share/opensearch/config/securityconfig-local:ro
    expose:
      - "9200"
      - "9600"
    networks:
      - opensearch_net

  dashboards:
    image: ${OPENSEARCH_DASHBOARDS_IMAGE}
    container_name: opensearch-dashboards
    restart: unless-stopped
    env_file:
      - .env
    depends_on:
      - opensearch
    environment:
      - TZ=${TZ}
      - http_proxy=${http_proxy}
      - https_proxy=${https_proxy}
      - no_proxy=${no_proxy}
      - HTTP_PROXY=${HTTP_PROXY}
      - HTTPS_PROXY=${HTTPS_PROXY}
      - NO_PROXY=${NO_PROXY}
    volumes:
      - ./dashboards/opensearch_dashboards.yml:/usr/share/opensearch-dashboards/config/opensearch_dashboards.yml:ro
    expose:
      - "5601"
    networks:
      - opensearch_net

  nginx:
    image: ${NGINX_IMAGE}
    container_name: opensearch-nginx
    restart: unless-stopped
    depends_on:
      - dashboards
    ports:
      - "80:80"
      - "8080:8080"
      - "443:443"
      - "8443:8443"
      - "9200:9200"
    volumes:
      - ./nginx/nginx.conf:/etc/nginx/nginx.conf:ro
      - ./certs:/etc/nginx/certs:ro
    networks:
      - opensearch_net

networks:
  opensearch_net:
    name: opensearch_network
```

Download the official ONNX model:
https://huggingface.co/onnx-community/bge-reranker-v2-m3-ONNX
Register it in ML Commons as a TEXT_SIMILARITY model using this script

```
import argparse
import os
import sys
import urllib3
from opensearchpy import OpenSearch
from opensearch_py_ml.ml_commons import MLCommonClient

# Desactivar advertencias de certificados auto-firmados
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)


def main():
    # Configuración de los argumentos por línea de comandos
    parser = argparse.ArgumentParser(
        description="Subir modelos pesados a OpenSearch ML Commons con chunking automático usando rutas relativas."
    )
    parser.add_argument(
        "--model_path",
        required=True,
        help="Ruta (puede ser relativa) al archivo .zip del modelo (ej: ./bge-m3.zip)",
    )
    parser.add_argument(
        "--model_config_path",
        required=True,
        help="Ruta (puede ser relativa) al archivo metadata_model.json",
    )
    parser.add_argument(
        "--model_group_id",
        required=True,
        help="ID del grupo de modelos en OpenSearch",
    )

    args = parser.parse_args()

    # Convertimos las rutas relativas que pases en la consola a rutas absolutas reales
    # para que Python las localice sin importar desde dónde ejecutes el script.
    resolved_model_path = os.path.abspath(args.model_path)
    resolved_config_path = os.path.abspath(args.model_config_path)

    # Validación rápida para avisarte si el archivo no existe antes de conectar con OpenSearch
    if not os.path.exists(resolved_model_path):
        print(
            f"Error: No se encontró el archivo del modelo en: {resolved_model_path}"
        )
        sys.exit(1)
    if not os.path.exists(resolved_config_path):
        print(
            f"Error: No se encontró el archivo de configuración en: {resolved_config_path}"
        )
        sys.exit(1)

    print(
        "Conectando con el cluster de OpenSearch desde el contenedor de Python..."
    )

    client = OpenSearch(
        hosts=[{"host": "172.18.0.2", "port": 9200}],
        http_auth=("admin", "admin"),
        use_ssl=True,
        verify_certs=False,
        ssl_show_warn=False,
         timeout=120,
         max_retries=5,
         retry_on_timeout=True
    )

    ml_client = MLCommonClient(client)

    print("\n-------------------------------------------------------------")
    print("Iniciando registro e inyección de chunks...")
    print(f"-> Archivo de modelo (resuelto): {resolved_model_path}")
    print(f"-> Configuración (resuelta): {resolved_config_path}")
    print(f"-> Grupo de modelos: {args.model_group_id}")
    print("-------------------------------------------------------------\n")

    try:
        model_id = ml_client.register_model(
            model_path=resolved_model_path,
            model_config_path=resolved_config_path,
            model_group_id=args.model_group_id,
            isVerbose=True,
            deploy_model=False,
        )
        print("\n-------------------------------------------------------------")
        print("¡ÉXITO TOTAL! El modelo fue registrado y subido.")
        print(f"ID del modelo generado: {model_id}")
        print("-------------------------------------------------------------")
        print("Próximo paso manual en Dev Tools o cURL:")
        print(f"POST /_plugins/_ml/models/{model_id}/_deploy")
        print("-------------------------------------------------------------")

    except Exception as e:
        print(f"\nError durante el registro: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

```
usage: 
python3 upload_model_docker.py --model_path "./bge-reranker-v2-m3-ONNX.zip" --model_config_path "./metadata_model.json" --model_group_id <any group>

bge-reranker-v2-m3-ONNX.zip contains 

config.json
model.onnx
special_tokens_map.json
tokenizer.json
tokenizer_config.json

Metadata used:

```
{
  "name": "bge-reranker-v2-m3-onnx",
  "version": "1.0.0",
  "model_format": "ONNX",
  "function_name": "TEXT_SIMILARITY",
  "model_config": {
    "model_type": "onnx",
    "framework_type": "sentence_transformers",
    "embedding_dimension": 1,
    "all_config": "{...original config.json...}"
  }
}
```
Deploy the model successfully.
Execute:
POST /_plugins/_ml/_predict/text_similarity/<model_id>

with

{
  "query_text": "The user completed the digital registration but cannot upload ID photos again.",
  "text_docs": [
    "asking a question."
  ]
}
The request always fails with:
{
  "error": {
    "type": "m_l_exception",
    "reason": "Failed to inference TEXT_SIMILARITY model",
    "caused_by": {
      "type": "translate_exception",
      "reason": "java.lang.IllegalArgumentException: Input mismatch, looking for: [input_ids, attention_mask]"
    }
  }
}

**What is the expected behavior?**
A ranking of text similarity

Is onnx-community/bge-reranker-v2-m3-ONNX expected to be supported by TEXT_SIMILARITY in OpenSearch 3.6, or are only specific ONNX cross-encoder models currently supported?

### PR DESCRIPTION
**Pull request by nathaliellenaa**
### Description
TEXT_SIMILARITY inference failed with `Input mismatch, looking for: [input_ids, attention_mask]` for HuggingFace ONNX cross-encoder models that declare only two inputs (`input_ids`, `attention_mask`), such as [onnx-community/bge-reranker-v2-m3-ONNX](https://huggingface.co/onnx-community/bge-reranker-v2-m3-ONNX). These models register and deploy successfully, but every prediction failed before model execution.

The root cause is in `TextSimilarityTranslator.processInput`: it unconditionally appended a third `token_type_ids` tensor, while many exported ONNX cross-encoders (e.g. XLM-RoBERTa–based rerankers) declare only `input_ids` and `attention_mask`. DJL rejects the extra input, so inference never runs.

This change makes `token_type_ids` conditional on the loaded model's actual input signature:
  - Inspect `ctx.getModel().getBlock().describeInput()` and only append the `token_type_ids` tensor when the model declares it.
  - If the signature is unavailable (null/empty) or cannot be inspected, fall back to the previous behavior (send `token_type_ids)`, so existing BERT-style models are unaffected.

**Behavior is a strict superset of the old behavior** — no model that worked before changes:
```
  ┌──────────────────────────────────────────┬───────────────────┬────────────────────────────────┐
  │          Model input signature           │      Before       │             After              │
  ├──────────────────────────────────────────┼───────────────────┼────────────────────────────────┤
  │ Declares token_type_ids (BERT-style)     │ 3 tensors ✅      │ 3 tensors ✅                   │
  ├──────────────────────────────────────────┼───────────────────┼────────────────────────────────┤
  │ Declares only input_ids + attention_mask │ Input mismatch ❌ │ 2 tensors ✅                   │
  ├──────────────────────────────────────────┼───────────────────┼────────────────────────────────┤
  │ Signature unavailable                    │ 3 tensors         │ 3 tensors (unchanged fallback) │
  └──────────────────────────────────────────┴───────────────────┴────────────────────────────────┘
```
No wire-format, SPI, or REST API changes — the fix is confined to ml-algorithms runtime inference logic, so it is backward-compatible for rolling upgrades.

### Related Issues
Resolves https://github.com/opensearch-project/ml-commons/issues/4924

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by mingshl** [comment]
> ## PR Code Analyzer ❗
> 
> **AI-powered '_Code-Diff-Analyzer_' found issues on commit [3051140](https://github.com/opensearch-project/ml-commons/commit/3051140a0d984d4aa179bbcbcd1ea65596e7a02d).**
> 
> ⛔ **Hard block**: Issues at **High** severity or above will block this PR from merging.
> Path	Line	Severity	Description
> ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_similarity/TinyXLMRoberta-CE-onnx.zip	1	medium	Binary ZIP file added as a test resource. Contents cannot be audited through code review. The file is loaded and executed during tests via initModel(); maintainers should verify it originates from a trusted source and has not been tampered with (e.g., confirm its SHA-256 hash against the canonical TinyXLMRoberta model artifact).
> 
> The table above displays the top 10 most important findings. **Total: 1 | Critical: 0 | High: 0 | Medium: 1 | Low: 0**
> 
> **Pull Requests Author(s)**: Please update your Pull Request according to the report above.
> 
> **Repository Maintainer(s)**: You can `bypass diff analyzer` by adding label `skip-diff-analyzer` after reviewing the changes carefully, then `re-run failed actions`. To re-enable the analyzer, remove the label, then `re-run all actions`.
> 
> ⚠️ Note: The _Code-Diff-Analyzer_ helps protect against potentially harmful code patterns. Please ensure you have thoroughly reviewed the changes beforehand.
> 
> Thanks.

@nathaliellenaa the zip file is not required in the PR right? would you remove the zip file.

**Comment by nathaliellenaa** [comment]
> > ## PR Code Analyzer ❗
> > **AI-powered '_Code-Diff-Analyzer_' found issues on commit [3051140](https://github.com/opensearch-project/ml-commons/commit/3051140a0d984d4aa179bbcbcd1ea65596e7a02d).**
> > ⛔ **Hard block**: Issues at **High** severity or above will block this PR from merging.
> > Path	Line	Severity	Description
> > ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_similarity/TinyXLMRoberta-CE-onnx.zip	1	medium	Binary ZIP file added as a test resource. Contents cannot be audited through code review. The file is loaded and executed during tests via initModel(); maintainers should verify it originates from a trusted source and has not been tampered with (e.g., confirm its SHA-256 hash against the canonical TinyXLMRoberta model artifact).
> > The table above displays the top 10 most important findings. **Total: 1 | Critical: 0 | High: 0 | Medium: 1 | Low: 0**
> > **Pull Requests Author(s)**: Please update your Pull Request according to the report above.
> > **Repository Maintainer(s)**: You can `bypass diff analyzer` by adding label `skip-diff-analyzer` after reviewing the changes carefully, then `re-run failed actions`. To re-enable the analyzer, remove the label, then `re-run all actions`.
> > ⚠️ Note: The _Code-Diff-Analyzer_ helps protect against potentially harmful code patterns. Please ensure you have thoroughly reviewed the changes beforehand.
> > Thanks.
> 
> @nathaliellenaa the zip file is not required in the PR right? would you remove the zip file.

@mingshl the zip is required for the test as currently written (it loads a real 2-input ONNX model through `initModel()`), but we can drop the zip and replace it with a mock-based unit test. This will cover the fix without shipping a binary.

Update: addressed this in the new commit

**Comment by mingshl** [comment]
> > > ## PR Code Analyzer ❗
> > > **AI-powered '_Code-Diff-Analyzer_' found issues on commit [3051140](https://github.com/opensearch-project/ml-commons/commit/3051140a0d984d4aa179bbcbcd1ea65596e7a02d).**
> > > ⛔ **Hard block**: Issues at **High** severity or above will block this PR from merging.
> > > Path	Line	Severity	Description
> > > ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_similarity/TinyXLMRoberta-CE-onnx.zip	1	medium	Binary ZIP file added as a test resource. Contents cannot be audited through code review. The file is loaded and executed during tests via initModel(); maintainers should verify it originates from a trusted source and has not been tampered with (e.g., confirm its SHA-256 hash against the canonical TinyXLMRoberta model artifact).
> > > The table above displays the top 10 most important findings. **Total: 1 | Critical: 0 | High: 0 | Medium: 1 | Low: 0**
> > > **Pull Requests Author(s)**: Please update your Pull Request according to the report above.
> > > **Repository Maintainer(s)**: You can `bypass diff analyzer` by adding label `skip-diff-analyzer` after reviewing the changes carefully, then `re-run failed actions`. To re-enable the analyzer, remove the label, then `re-run all actions`.
> > > ⚠️ Note: The _Code-Diff-Analyzer_ helps protect against potentially harmful code patterns. Please ensure you have thoroughly reviewed the changes beforehand.
> > > Thanks.
> > 
> > 
> > @nathaliellenaa the zip file is not required in the PR right? would you remove the zip file.
> 
> @mingshl the zip is required for the test as currently written (it loads a real 2-input ONNX model through `initModel()`), but we can drop the zip and replace it with a mock-based unit test. This will cover the fix without shipping a binary.
> 
> Update: addressed this in the new commit

if it's just test purpose, mock is a better approach. The zip file of ml-commons is already large and we try to avoid a plugin going too big in file size.

**Review by mingshl — APPROVED** [review]

**Comment by mingshl** [comment]
should we backport to earlier version too? can you lookup if earlier version is impacted?

**Comment by nathaliellenaa** [comment]
> should we backport to earlier version too? can you lookup if earlier version is impacted?

I believe the behavior has been present since the feature was introduced. I reproduced it on 2.x - 3.7 versions, they are affected.

**Comment by llocani** [comment]
Hello ¿Is there any branch with this change for v3.6? thanks in advance

**Comment by ylwu-amzn** [comment]
> > should we backport to earlier version too? can you lookup if earlier version is impacted?
> 
> I believe the behavior has been present since the feature was introduced. I reproduced it on 2.x - 3.7 versions, they are affected.

@nathaliellenaa Can you test to confirm? If this bug exists in previous versions, we should backport

**Comment by nathaliellenaa** [comment]
> > > should we backport to earlier version too? can you lookup if earlier version is impacted?
> > 
> > 
> > I believe the behavior has been present since the feature was introduced. I reproduced it on 2.x - 3.7 versions, they are affected.
> 
> @nathaliellenaa Can you test to confirm? If this bug exists in previous versions, we should backport

Confirmed this bug exists since the feature first introduced in 2.12. I will raise backport PRs to all affected versions
